### PR TITLE
Add a CPU utilization resource monitor for overload manager

### PR DESCRIPTION
**Pull request by cancecen**
Commit Message: Add a CPU utilization resource monitor for overload manager
Additional Description: Adds a new resource monitor to be used in CPU bound workloads to shed load. i.e. this can be configured to reject requests once CPU Utilization reaches a certain brownout point. 

In my company, we experience user driven retry storms and/or unexpected flash crowds that are mitigated via auto scaling of our compute resources. However, auto scaling takes time and we still want to keep our fleet reasonably healthy during auto scaling and having a way to cheaply reject requests via overload manager has been very helpful. As a platform owner, it is easier to configure limits for CPU bound workloads with this resource monitor than other RPS/Latency/Concurrency based limiters. 

Risk Level: Low
Testing: Unit tests
Docs Changes: 
Release Notes:
Platform Specific Features: Today this is only implemented for Linux

### CODE REVIEW

File: source/extensions/resource_monitors/cpu_utilization/config.cc [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace Envoy {
+namespace Extensions {
+namespace ResourceMonitors {
+namespace CpuUtilizationMonitor {
+
+Server::ResourceMonitorPtr CpuUtilizationMonitorFactory::createResourceMonitorFromProtoTyped(
+    const envoy::extensions::resource_monitors::cpu_utilization::v3::CpuUtilizationConfig& config,
+    Server::Configuration::ResourceMonitorFactoryContext& /*unused_context*/) {
+  // In the future, the below can be configurable based on the operating system.
```
Comment (KBaichoo) [inline source/extensions/resource_monitors/cpu_utilization/config.cc:19]: Can we make it not compiled in for other unsupported targets?

File: source/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace Envoy {
+namespace Extensions {
+namespace ResourceMonitors {
+namespace CpuUtilizationMonitor {
+
+static const std::string LINUX_CPU_STATS_FILE = "/proc/stat";
+
+constexpr uint64_t NUMBER_OF_CPU_TIMES_TO_PARSE = 4;
+constexpr uint64_t MICROSECONDS = 1000 * 1000;
```
Comment (KBaichoo) [inline source/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader.h:15]: These can go in the implementation file? Also can NUMER_OF_CPU_TIMES_TO_PARSE be commented? Why 4 times?

File: source/extensions/resource_monitors/cpu_utilization/cpu_utilization_monitor.cc [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        CpuUtilizationConfig& /*config*/,
+    std::unique_ptr<CpuStatsReader> cpu_stats_reader)
+    : utilization_(0.0), cpu_stats_reader_(std::move(cpu_stats_reader)) {
+  previous_cpu_times_ = cpu_stats_reader_->getCpuTimes();
+}
+
+void CpuUtilizationMonitor::updateResourceUsage(Server::ResourceUpdateCallbacks& callbacks) {
+  CpuTimes cpu_times = cpu_stats_reader_->getCpuTimes();
+  if (!cpu_times.is_valid) {
+    ENVOY_LOG_MISC(error, "Can't open file to read CPU utilization");
```
Comment (KBaichoo) [inline source/extensions/resource_monitors/cpu_utilization/cpu_utilization_monitor.cc:32]: I think just putting the rrors in the onFailure call should be sufficient enough

File: source/extensions/resource_monitors/cpu_utilization/cpu_utilization_monitor.cc [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    const auto& error = EnvoyException("Can't open file to read CPU utilization");
+    callbacks.onFailure(error);
+    return;
+  }
+
+  const int64_t work_over_period = cpu_times.work_time - previous_cpu_times_.work_time;
+  const int64_t total_over_period = cpu_times.total_time - previous_cpu_times_.total_time;
+  if (work_over_period < 0 || total_over_period <= 0) {
+    ENVOY_LOG_MISC(error,
```
Comment (KBaichoo) [inline source/extensions/resource_monitors/cpu_utilization/cpu_utilization_monitor.cc:41]: likewise as above -- there's a failed update stat that gets incremented that can be used for monitoring the failures

File: source/extensions/resource_monitors/cpu_utilization/cpu_utilization_monitor.cc [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  }
+  const double current_utilization = static_cast<double>(work_over_period) / total_over_period;
+  ENVOY_LOG_MISC(trace, "Prev work={}, Cur work={}, Prev Total={}, Cur Total={}",
+                 previous_cpu_times_.work_time, cpu_times.work_time, previous_cpu_times_.total_time,
+                 cpu_times.total_time);
+  // The new utilization is calculated/smoothed using EWMA
+  // (https://en.wikipedia.org/wiki/EWMA_chart) where we choose a very small dampening alpha to give
+  // past calculations higher priority.
+  utilization_ = current_utilization * DAMPENING_ALPHA + (1 - DAMPENING_ALPHA) * utilization_;
+  ENVOY_LOG_MISC(trace, "CPU Utilization={}", utilization_ * 100);
```
Comment (KBaichoo) [inline source/extensions/resource_monitors/cpu_utilization/cpu_utilization_monitor.cc:57]: minor: onSuccess will update the pressure gauge stat with this utilization which might be sufficient for visibility vs logging it.

File: test/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader_test.cc [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace ResourceMonitors {
+namespace CpuUtilizationMonitor {
+namespace {
+
+TEST(CpuStatsReader, ReadsCpuStatsEc2) {
```
Comment (KBaichoo) [inline test/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader_test.cc:18]: Ec2 leakage -- not dependent on Ec2.

File: test/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader_test.cc [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  file_updater.update(contents);
+
+  LinuxCpuStatsReader cpu_stats_reader(temp_path);
+  CpuTimes cpu_times = cpu_stats_reader.getCpuTimes();
+
+  EXPECT_EQ(cpu_times.work_time, 17995597);
+  EXPECT_EQ(cpu_times.total_time, 29590585);
+}
+
+TEST(CpuStatsReader, CannotReadFileEc2) {
```
Comment (KBaichoo) [inline test/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader_test.cc:48]: likewise

**Review by KBaichoo — COMMENTED** [review]
Thank you for working on this! 

/wait

File: test/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader_test.cc (hunk 6 above)
Comment (cancecen) [reply test/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader_test.cc:18]: Fixed

File: test/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader_test.cc (hunk 7 above)
Comment (cancecen) [reply test/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader_test.cc:48]: Fixed

File: source/extensions/resource_monitors/cpu_utilization/cpu_utilization_monitor.cc (hunk 3 above)
Comment (cancecen) [reply source/extensions/resource_monitors/cpu_utilization/cpu_utilization_monitor.cc:32]: Yeah, makes sense. Removed.

File: source/extensions/resource_monitors/cpu_utilization/cpu_utilization_monitor.cc (hunk 4 above)
Comment (cancecen) [reply source/extensions/resource_monitors/cpu_utilization/cpu_utilization_monitor.cc:41]: Removed.

File: source/extensions/resource_monitors/cpu_utilization/cpu_utilization_monitor.cc (hunk 5 above)
Comment (cancecen) [reply source/extensions/resource_monitors/cpu_utilization/cpu_utilization_monitor.cc:57]: Removed.

File: api/envoy/extensions/resource_monitors/cpu_utilization/v3/cpu_utilization.proto
[the remark this answers is not in the document]
Comment (cancecen) [reply api/envoy/extensions/resource_monitors/cpu_utilization/v3/cpu_utilization.proto:1]: Added documentation.

File: source/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader.h (hunk 2 above)
Comment (cancecen) [reply source/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader.h:15]: Moved.

File: source/extensions/resource_monitors/cpu_utilization/cpu_utilization_monitor.h [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+#include "source/common/runtime/runtime_features.h"
+#include "source/extensions/resource_monitors/cpu_utilization/cpu_stats_reader.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace ResourceMonitors {
+namespace CpuUtilizationMonitor {
+
+constexpr double DAMPENING_ALPHA = 0.05;
```
[the remark this answers is not in the document]
Comment (cancecen) [reply source/extensions/resource_monitors/cpu_utilization/cpu_utilization_monitor.h:16]: Added a comment and moved to cc.

**Comment by KBaichoo** [comment]
Some other ideas: We use cgroups cpu controller to limit CPU usage in a container environment. So maybe some cgroup metrics, eg. [cpu.stat](https://kernel.googlesource.com/pub/scm/linux/kernel/git/glommer/memcg/+/cpu_stat/Documentation/cgroups/cpu.txt#:~:text=%2D-,cpu.stat,-%3A%20statistics%20about%20the) is a valuable resource monitor for overload manager.
> @labilezhu agree this would be a good follow up to have a different source for the cpu monitor

/wait

File: docs/root/configuration/operations/overload_manager/overload_manager.rst [hunk 9]
```diff
@@ -356,34 +356,34 @@ there's something seriously wrong e.g. in this example streams using ``>=
...
-       typed_config:
-         "@type": type.googleapis.com/envoy.extensions.resource_monitors.cpu_utilization.v3.CpuUtilizationConfig
-   actions:
-     - name: "envoy.overload_actions.stop_accepting_requests"
-       triggers:
-         - name: "envoy.resource_monitors.cpu_utilization"
-           scaled:
-             scaling_threshold: 0.80
-             saturation_threshold: 0.95
+  refresh_interval:
```
Comment (phlax) [inline docs/root/configuration/operations/overload_manager/overload_manager.rst:373]: would it be possible to move this to a boostrap config and `literalinclude` it ?

it just means it gets tested and is more useful for end users

Comment (cancecen) [reply docs/root/configuration/operations/overload_manager/overload_manager.rst:373]: Can you please share an example? I am not sure if I fully understand what you mean. I just followed the convention from this particular file and would like to follow if there is a better way of doing it.

File: source/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader.cc [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+CpuTimes LinuxCpuStatsReader::getCpuTimes() {
+  std::ifstream cpu_stats_file;
+  cpu_stats_file.open(cpu_stats_filename_);
+  if (!cpu_stats_file.is_open()) {
+    ENVOY_LOG_MISC(error, "Can't open ec2 cpu stats file {}", cpu_stats_filename_);
+    return {false, 0, 0};
+  }
+
+  cpu_stats_file.ignore(5, ' '); // Skip the 'cpu' prefix.
```
[the remark this answers is not in the document]
Comment (cancecen) [reply source/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader.cc:23]: Addressed. I validated `cpu` aggregate line should come first and also we need to have at least 4 tokens.

File: source/extensions/resource_monitors/cpu_utilization/config.cc (hunk 1 above)
Comment (cancecen) [reply source/extensions/resource_monitors/cpu_utilization/config.cc:19]: Done

File: docs/root/configuration/operations/overload_manager/overload_manager.rst (hunk 9 above)
Comment (phlax) [reply docs/root/configuration/operations/overload_manager/overload_manager.rst:373]: sorry for slow reply - i missed this comment

you can find existing examples with

```console
$ git grep literalinclude docs/
```

some examples provide a captioned download - which i think is generally preferable

**Comment by cancecen** [comment]
> /docs

Thanks @phlax. Looks good to me, let me know what you think:
https://storage.googleapis.com/envoy-pr/2eeb3fe/docs/configuration/operations/overload_manager/overload_manager.html#cpu-intensive-workload-brownout-protection

**Review by phlax — DISMISSED** [review]
nice! thanks

lgtm from docs pov - thanks for iterating

File: source/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader.cc [hunk 11]
```diff
@@ -20,10 +22,22 @@ CpuTimes LinuxCpuStatsReader::getCpuTimes() {
     return {false, 0, 0};
   }
 
-  cpu_stats_file.ignore(5, ' '); // Skip the 'cpu' prefix.
+  // The first 5 bytes should be 'cpu ' without a cpu index.
+  std::string buffer(5, '\0');
+  cpu_stats_file.read(buffer.data(), 5);
+  const std::string target = "cpu  ";
+  if (!cpu_stats_file || !std::equal(buffer.begin(), buffer.end(), target.begin(), target.end())) {
```
Comment (KBaichoo) [inline source/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader.cc:29]: you can just compare `buffer != target`

**Comment by kyessenov** [comment]
Somewhat related question: I see that you use kernel reported CPU usage here, but we use self-reported memory usage in `fixed_heap`. If we were consistent, we'd probably use cgroup2 `/sys/fs/cgroup/memory.current` for memory monitor. Is there some rationale why we choose kernel here but user-space monitor there?

**Comment by cancecen** [comment]
> Somewhat related question: I see that you use kernel reported CPU usage here, but we use self-reported memory usage in `fixed_heap`. If we were consistent, we'd probably use cgroup2 `/sys/fs/cgroup/memory.current` for memory monitor. Is there some rationale why we choose kernel here but user-space monitor there?

Yes. We use the kernel reported one because we really want to get the view of the whole system - i.e. there can be workloads Envoy is proxying requests running in the system that are actually compute intensive. This way we can make the decision based on that.

**Comment by kyessenov** [comment]
> > Somewhat related question: I see that you use kernel reported CPU usage here, but we use self-reported memory usage in `fixed_heap`. If we were consistent, we'd probably use cgroup2 `/sys/fs/cgroup/memory.current` for memory monitor. Is there some rationale why we choose kernel here but user-space monitor there?
> 
> Yes. We use the kernel reported one because we really want to get the view of the whole system - i.e. there can be workloads Envoy is proxying requests running in the system that are actually compute intensive. This way we can make the decision based on that.

Is there a reason why you are not using `cpu.stat` from `cgroup2`? That would work better on k8s, for example, where Envoy runs inside a container.

**Comment by cancecen** [comment]
> > > Somewhat related question: I see that you use kernel reported CPU usage here, but we use self-reported memory usage in `fixed_heap`. If we were consistent, we'd probably use cgroup2 `/sys/fs/cgroup/memory.current` for memory monitor. Is there some rationale why we choose kernel here but user-space monitor there?
> > 
> > 
> > Yes. We use the kernel reported one because we really want to get the view of the whole system - i.e. there can be workloads Envoy is proxying requests running in the system that are actually compute intensive. This way we can make the decision based on that.
> 
> Is there a reason why you are not using `cpu.stat` from `cgroup2`? That would work better on k8s, for example, where Envoy runs inside a container.

That can be a follow up to this PR. Today in my company we have 3 different compute platforms and the majority of the workloads are in VMs. That's why I structured the code to have the possibility of separate cpu stats reader per platform.

**Review by KBaichoo — APPROVED** [review]

**Pull request merged**